design/ceph/block-ceph-delete-if-pool-data.md

```diff
@@ -0,0 +1,146 @@
+# Blocking CephCluster deletion if user data is present
+
+
+## Use case
+As an administrator, I want to be protected from accidentally deleting user data if the CephCluster
+is deleted (accidentally or on purpose).
+
+
+## Determining if a Ceph is in use
```

Comment:
After reading through this and pondering all our discussions in the last week, this is where I'm landing...
1. Each controller that owns creating/deleting Ceph resources would block removal if those resources appear to contain data
2. Each controller that has dependent controllers would block removal if any dependent CRs exist.

For example, this would translate into:
- CephCluster: block if any dependent CRs exist in its own namespace: CephBlockPool, CephObjectStore, CephFilesystem, CephNFS, ...
- CephObjectStore: block if any OBs exist (already implemented), or if any CephNFS CRs are referencing the object store
- CephBlockPool: block if any rbd images exist in the pool (already implemented)
- CephFilesystem: block if the filesystem appears in use, or if any CephNFS CRs are referencing the filesystem 

Of note is that the CephCluster doesn't directly check for any Ceph resources. By checking for dependent CRs which in turn own Ceph resources, we effectively block since there might be Ceph resources in use. If the user deletes those dependent CRs, then the CephCluster would be able to proceed with the removal.

I believe this keeps the implementation simple and also accomplishes the goals of 1) blocking uninstall when there is data and 2) not overly complicating dev scenarios.

Reply:
Would you also  include not deleting RBD mirroring CRs if block pools reference them?
Or deleting objcet realm/zonegroup/zone if they are referenced by objectstores?

CephClients can also reference pools and could be dependents of Block/File/Object resources that create pools just as NFS can.

I think that we should probably implement dependent safety checks for all Rook resources if we are going to do them for some so that users aren't confused if they aren't protected in some cases when they are in others and so we don't have to elaborately document which cases are and which cases are not protected.

Reply:
> Would you also include not deleting RBD mirroring CRs if block pools reference them?
> Or deleting objcet realm/zonegroup/zone if they are referenced by objectstores?

I was just providing examples, agreed there are other safety checks to add as you mentioned.

> CephClients can also reference pools and could be dependents of Block/File/Object resources that create pools just as NFS can.

While CephClients may be used for access to those resources, I don't see it as critical to detect. It's a weaker association that doesn't affect the ability to preserve data. Or at least there is a workaround to use the admin key if needed.

> I think that we should probably implement dependent safety checks for all Rook resources if we are going to do them for some so that users aren't confused if they aren't protected in some cases when they are in others and so we don't have to elaborately document which cases are and which cases are not protected.

Certainly, let's just agree on the goal. If it's potential data loss, we should protect against it. If there is some weaker association or there is a workaround, it may not be necessary.